porespy/generators/__imgen__.py

```diff
@@ -974,30 +975,157 @@ def cylinders(shape: List[int], radius: int, ncylinders: int,
+        lie *out of* the XY plane, with 0 meaning all cylinders lie in the XY
+        plane, and 90 meaning that cylinders are randomly oriented out of the
+        plane by as much as +/- 90 degrees.
+    theta_max : scalar
+        A value between 0 and 90 that controls the amount of rotation *in the*
+        XY plane, with 0 meaning all cylinders point in the X-direction, and
+        90 meaning they are randomly rotated about the Z axis by as much
+        as +/- 90 degrees.
+    length : scalar
+        The length of the cylinders to add.  If ``None`` (default) then the
+        cylinders will extend beyond the domain in both directions so no ends
+        will exist. If a scalar value is given it will be interpreted as the
+        Euclidean distance between the two ends of the cylinder.  Note that
+        one or both of the ends *may* still lie outside the domain, depending
+        on the randomly chosen center point of the cylinder.
+    max_iter : scalar
+        The number of fractional fiber insertions used to target the requested
+        porosity. By default a value of 3 is used (and this is typically
+        effective in getting very close to the targeted porosity), but a
+        greater number can be input to improve the achieved porosity.
+    return_fiber_number : bool
+        Determines whether the function will return the number of fibers
+        along with the image
+
+    Returns
+    -------
+    image : ND-array
+        A boolean array with ``True`` values denoting the pore space
+    fiber_number : scalar
+        The number of fibers added to the image. This is optional, and whether
+        it is returned is determined by 'return_fiber_number'
+    """
+    if max_iter < 3:
+        raise Exception("Iterations must be greater than or equal to 3")
+
+    vol_total = float(np.prod(shape))
+
+    def get_num_pixels(porosity):
+        r"""
+        Helper method to calculate number of pixels given a porosity
+        """
+        return -np.log(porosity) * vol_total
+
+    # Crudely estimate fiber length as cube root of product of dims
+    length_estimate = vol_total ** (1 / 3) if length is None else length
+
+    # Rough fiber volume estimate
+    vol_fiber = length_estimate * np.pi * radius * radius
+    n_pixels_to_add = get_num_pixels(porosity)
+
+    # Rough estimate of n_fibers
+    n_fibers_added = 0
+    # Calculate fraction of fibers to be added in each iteration.
+    subdif = 0.8 / np.sum(np.arange(1, max_iter) ** 2)
+    fractions = [0.2]
+    for i in range(1, max_iter):
+        fractions.append(fractions[i - 1] + (max_iter - i) ** 2 * subdif)
+
+    im = np.ones(shape, dtype=bool)
+    for frac in tqdm(fractions, disable=not verbose, file=sys.stdout,
+                     desc="Adding fibers"):
+        n_fibers_total = n_pixels_to_add / vol_fiber
+        n_fibers = int(np.ceil(frac * n_fibers_total) - n_fibers_added)
+        if n_fibers > 0:
+            im = im & ps.generators.cylinders(
```

Comment:
I'm inclined to think that we should make the current 'cylinders' function a hidden method (_cylinders) and make this new one the official 'cylinders' method, and add only the `porosity` to the func call.  If `porosity=None` and `n_cylinders=None` are the defaults, then the user can specify one or the other, and so both new and old behaviors can be obtained.

Reply:
Good point, done